File: DiscoDruid3/app/src/main/java/com/andruid/magic/discodruid/viewmodel/PlaylistViewModel.java
package com.andruid.magic.discodruid.viewmodel;

import android.app.Application;
import android.support.v4.media.MediaBrowserCompat;

import com.andruid.magic.discodruid.data.Constants;
import com.andruid.magic.discodruid.datasourcefactory.PlaylistDataSourceFactory;
import com.andruid.magic.discodruid.model.PlayList;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.paging.LivePagedListBuilder;
import androidx.paging.PagedList;

public class PlaylistViewModel extends AndroidViewModel {

    private Application application;

    public PlaylistViewModel(@NonNull Application application) {
        super(application);
        this.application = application;
    }

    public LiveData<PagedList<PlayList>> getPlayLists(MediaBrowserCompat mediaBrowserCompat){
        PlaylistDataSourceFactory playlistDataSourceFactory = new PlaylistDataSourceFactory(mediaBrowserCompat, application.getApplicationContext());
        PagedList.Config config = new PagedList.Config.Builder()
                .setEnablePlaceholders(false)
                .setPageSize(Constants.PAGE_SIZE)
                .build();
        return new LivePagedListBuilder<>(playlistDataSourceFactory,config).build();
    }
}

File: DiscoDruid/app/src/main/java/com/andruid/magic/discodruid/ui/viewmodel/PlaylistViewModel.kt
package com.andruid.magic.discodruid.ui.viewmodel

import android.support.v4.media.MediaBrowserCompat
import androidx.lifecycle.ViewModel
import androidx.lifecycle.asLiveData
import androidx.lifecycle.liveData
import androidx.lifecycle.viewModelScope
import androidx.paging.Pager
import androidx.paging.PagingConfig
import androidx.paging.cachedIn
import com.andruid.magic.discodruid.data.PAGE_SIZE
import com.andruid.magic.discodruid.paging.PlaylistsPagingSource
import kotlinx.coroutines.cancel

class PlaylistViewModel(private val mediaBrowserCompat: MediaBrowserCompat) : ViewModel() {
    val playlistLiveData = liveData {
        val config = PagingConfig(PAGE_SIZE)
        val pager = Pager(config) {
            PlaylistsPagingSource(mediaBrowserCompat)
        }

        emitSource(
            pager.flow.cachedIn(viewModelScope)
                .asLiveData()
        )
    }

    override fun onCleared() {
        super.onCleared()
        viewModelScope.cancel()
    }
}